routes/api.js

```diff
@@ -40,164 +36,50 @@ var proxy = httpProxy.createProxyServer({
     protocolRewrite: true
 });
 
-var substituteCodeCell = function(payload) {
-    debug('PROXY: received message from client WS: ' + (payload));
-    var transformedData = payload;
-
-    // substitute in code if necessary
-    // for performance reasons, first do a quick string check before JSON parsing
-    if (payload.indexOf('execute_request') !== -1) {
-        try {
-            payload = JSON.parse(payload);
-            if (payload.header.msg_type === 'execute_request') {
-                // get notebook data for current session
-                var nbpath = urlToDashboard(sessions[payload.header.session]);
-                transformedData = nbstore.get(nbpath).then(
-                    function success(nb) {
-                        // get code string for cell at index and update WS message
-
-                        // code must be an integer corresponding to a cell index
-                        var cellIdx = parseInt(payload.content.code, 10);
-
-                        // code must only consist of a non-negative integer
-                        if (cellIdx.toString(10) === payload.content.code &&
-                                cellIdx >= 0) {
-
-                            // substitute cell's actual code into the message
-                            var code = nb.cells[cellIdx].source.join('');
-                            payload.content.code = code;
-                            payload = JSON.stringify(payload);
-                        } else {
-                            // throw away execute request that has non-integer code
-                            payload = '';
-                        }
-                        return payload;
-                    },
-                    function failure(err) {
-                        error('Failed to load notebook data for', nbpath, err);
-                        return ''; // throw away execute request
-                    });
-            }
-        } catch(e) {
-            // TODO better handle parse error in WS message
-            error('Failed to parse `data` in WS.', e);
-            transformedData = '';
-        }
+function initWsProxy(server) {
+    if (wsProxy) {
```

Comment:
I tested with multiple clients and studied the code a bit more after I commented. I believe you're right that the only singleton is the WS server that accepts connections and spawns of instances of client/server to handle them independently. I couldn't see in the code any state that would be incorrectly shared across requests nor could I find a case in testing that causes crosstalk across clients. 👍